source/framework/vx_producer.c

```diff
@@ -0,0 +1,1336 @@
+
+    for (uint32_t i = 0; i < OVXGW_NUM_CLIENTS; i++)
+    {
+        if (producer->consumers_list[i].state == PROD_STATE_CLI_NOT_CONNECTED)
+        {
+            client = &producer->consumers_list[i];
+
+            client->state       = PROD_STATE_CLI_CONNECTED;
+            client->consumer_id = consumer_id;
+            client->socket_fd   = connection->socket_fd;
+            client_num          = i;
+
+            producer->nb_consumers++;
+            break;
+        }
+    }
+
+    if ((NULL == client) || (client_num == -1))
+    {
+        VX_PRINT(
+            VX_ZONE_ERROR, "PRODUCER %s: Maximum number of clients reached or error in client state\n", producer->name);
+    }
+
+    pthread_mutex_unlock(&producer->client_mutex);
+    return client_num;
+}
+
+static void drop_client(vx_producer producer, producer_bckchannel_t* client, int32_t client_num)
+{
+    pthread_mutex_lock(&producer->client_mutex);
+
+    VX_PRINT(VX_ZONE_INFO, "PRODUCER: Cleaning up client with socket %d and PID %d\n", client->socket_fd, client->consumer_id);
+
+    if (client->state != PROD_STATE_CLI_NOT_CONNECTED)
+    {
+        // zero out client info
+        client->state                 = PROD_STATE_CLI_NOT_CONNECTED;
+        client->first_buffer_released = 0;
+        client->consumer_id                   = 0;
+        client->socket_fd             = 0;
+        producer->nb_consumers--;
+
+        for (uint32_t i = 0; i < producer->numBufferRefs; i++)
+        {
+            // the ref we want to unlock will always be locked here, since the consumer is disconnected
+            // (at least one refcount is > 0 for locked)
+            if ((producer->refs[i].buffer_status == LOCKED) && (producer->refs[i].attached_to_client[client_num] == 1))
+            {
+                VX_PRINT(
+                    VX_ZONE_WARNING,
+                    "PRODUCER %s: Reference %u, was in LOCKED state, trying to FREE \n",
+                    producer->name,
+                    i,
+                    producer->refs[i].buffer_status);
+                producer->refs[i].attached_to_client[client_num] = 0;
+                int32_t status = set_buffer_status(producer->refs[i].ovx_ref, FREE, producer);
+                if (status != VXGW_STATUS_SUCCESS)
+                {
+                    VX_PRINT(
+                        VX_ZONE_ERROR, "PRODUCER %s: Reference %u, could not be set to FREE \n", producer->name, i);
+                }
+            }
+        }
+    }
+
+    pthread_mutex_unlock(&producer->client_mutex);
+
+    return;
+}
+
+static void handle_clients(void* clientPtr, void* data)
+{
+    // this function must be MT-safe
+    client_context* client   = (client_context*)clientPtr;
+    vx_producer producer = (vx_producer)data;
+
+    uint8_t           message_buffer[SOCKET_MAX_MSG_SIZE];
+    vx_gw_hello_msg* consumer_message;
+
+    int32_t client_num = -1;
+    int32_t status     = VXGW_STATUS_SUCCESS;
+
+    while (1)
```

Comment:
change the 1 by true? Or another variable to handle the shutdown?

Reply:
Reviewed and agreed